src/client/datascience/notebook/intellisense/pythonKernelCompletionProvider.ts

```diff
@@ -0,0 +1,322 @@
+                    // assume the current word needs to be replaced.
+                    // E.g. if you type in `os.env` and get complications from jupyter as `os.environ`, then
+                    // vscode will replace `env` with `os.environ`, as it replaces the word.
+                    // Leaving comment here so we know whats going on.
+                    // We cannot hardcode anything without any knowledge of what we're getting back.
+                };
+                return completion;
+            });
+        }
+
+        // Filter the list based on where we are in a cell (and the type of cell)
+        return filterCompletions(
+            context.triggerCharacter,
+            this.allowStringFilter,
+            completions,
+            pylanceResults,
+            document,
+            position
+        );
+    }
+    public async getJupyterCompletion(
+        session: IJupyterSession,
+        cellCode: string,
+        offsetInCode: number,
+        cancelToken?: CancellationToken
+    ): Promise<INotebookCompletion> {
+        // If server is busy, then don't delay code completion.
+        if (session.status === 'busy') {
+            return {
+                matches: [],
+                cursor: { start: 0, end: 0 },
+                metadata: {}
+            };
+        }
+        const result = await Promise.race([
+            session.requestComplete({
+                code: cellCode,
+                cursor_pos: offsetInCode
+            }),
+            createPromiseFromCancellation({ defaultValue: undefined, cancelAction: 'resolve', token: cancelToken })
+        ]);
+        traceInfoIfCI(
+            `Got jupyter notebook completions. Is cancel? ${cancelToken?.isCancellationRequested}: ${
+                result ? JSON.stringify(result) : 'empty'
+            }`
+        );
+        if (result && result.content) {
+            if ('matches' in result.content) {
+                return {
+                    matches: result.content.matches,
+                    cursor: {
+                        start: result.content.cursor_start,
+                        end: result.content.cursor_end
+                    },
+                    metadata: result.content.metadata
+                };
+            }
+        }
+        return {
+            matches: [],
+            cursor: { start: 0, end: 0 },
+            metadata: {}
+        };
+    }
+
+    private async getPylanceCompletions(
+        document: TextDocument,
+        position: Position,
+        context: CompletionContext,
+        cancelToken: CancellationToken
+    ) {
+        if (document.notebook) {
+            const client = await this.languageClientProvider.getLanguageClient(document.notebook);
+            if (client) {
+                // Use provider so it gets translated by middleware
+                const feature = client.getFeature(lsp.CompletionRequest.method);
+                const provider = feature.getProvider(document);
```

Comment:
This line here would fail if the document selector didn't match. Meaning Julia wouldn't go through here.